### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] Inconsistency and/or false positives with template literals

### ISSUE THREAD

**Issue by joshkel** [typescript-eslint/typescript-eslint#12276]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=C4TwDgpgBAaghgGwK4QAp2ACygXigcgDdEUBnfKAHygANjkJSA6AEgG9TgAnASwDsA5gF8aAbgBQ4gCYQAxgjhdoAMyR9ZwHgHs%2BUOGDAIQACjAZMpAFywSacwG0AugEprhLTykTxEAB5gtLmAoVXVNHSgkMCkMCGNPKyhOXkEnZyg2cSgoAHocqAB5AGs4EGtgUlkoJVktAT4eAC9GKCxoYAgAW0NYqAQeDq5ELL0DI2N7OltmdgT7AAZHERdJbLyoADE4HgRSKH4oABVwCABlWV4wYHK9gCIELVJGW9auno6%2BgYghhFaTvT4Un2fGU3z2yX4AhGtT4nCg9BQAEZcLQEYxWGw5osRBJsvpDCZ7GjES5vGt8sVSuUTpVLsAALSMfp8YJwWSyCBXPZwUhPILaXQ8qB8RRcLQAd0hrUw7RO0J0cLRACYUVMGDNMVJSAsljQ9Ht4Ax0FhcaMCRNlaTVrl8lsdnsDqBILSeFdGaRmdd9XzwrolAArOQdIEQQgQXRYLRIATYAZ9LRaIp7NpJOCdaBCuAAIy0YYA-CN8eMEkxOvpjPEpOkcAA%2BVHTDGeET6mxG8zOZwSITiIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
type ValuePath = 'values' | `values.${string}`;

declare function apply(paths: ValuePath[]): void;

export function update(ids: string[]) {
  // Okay: tsc recognizes the template literal
  apply([`values.${ids[0]}`])

  // Fails in TypeScript: ts "loses" template literal type and infers string
  const value1 = `values.${ids[0]}`;
  apply([value1]);

  // Okay: typescript-eslint accepts assertion as narrowing the type
  const value2 = `values.${ids[0]}` as ValuePath;
  apply([value2]);

  // Fails in typescript-eslint: assertion rejected even though it looks the same as above?
  apply(ids.map((id) => `values.${id}` as ValuePath));
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect the fourth example to be accepted by typescript-eslint, since the third example is and since the assertion is necessary for the code to typecheck.

### Actual Result

The fourth example generates an error:

> This assertion is unnecessary since the receiver accepts the original type of the expression.

**Comment by bradzacher** [comment]
For template literals you should use `as const`.

**Comment by joshkel** [comment]
Thanks, @bradzacher. It still seems to me that there's a potential inconsistency in how typescript-eslint 8.59 handles my third and fourth examples (<code>const value2 = \`values.${ids[0]}\` as ValuePath</code> is accepted, while <code>apply(ids.map((id) => \`values.${id}\` as ValuePath))</code> is rejected). But you're right that `as const` is better. (Feel free to disregard this if either I've misunderstood the potential inconsistency or if it's not a concern.)

**Comment by bradzacher** [comment]
Literals are really contextual. There's lot of special cases where literals are preserved and others where they widen.

In the case of a variable decl - the variable has no type, so the rule sees the assertion as necessary contextually. In the case of the map fn - the rule sees the function return type is inferred to be the literal type based off of the function argument type, so contextually it sees it as unnecessary. It's wrong in that second case though because of the weird rules of literals.

**Comment by bradzacher** [comment]
The rule probably shouldn't report on template literal expressions - I think the times when a template literal expression is inferred to be a literal type are pretty rare and so we should either encode those, explicitly error saying to use `as const` or ignore them.
